libraries/chain/include/eosio/chain/block_handle.hpp

```diff
@@ -0,0 +1,33 @@
+#pragma once
+
+#include <eosio/chain/block_state_legacy.hpp>
+#include <eosio/chain/block_state.hpp>
+
+namespace eosio::chain {
+
+// Created via controller::create_block_handle(const block_id_type& id, const signed_block_ptr& b)
+// Valid to request id and signed_block_ptr it was created from.
+struct block_handle {
```

Comment:
I think I would have made `struct block_handle : public std::variant<block_state_legacy_ptr, block_state_ptr>`, so we wouldn't need `internal()` for std::visit and other members like `block_num()`.